core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala

```diff
@@ -0,0 +1,432 @@
+      }
+
+      val ptr = WeakReference(ret)
+
+      var lastValue: Option[T] = None
+      val semaphore = new Semaphore(0)
+      val thread = new Thread(new Runnable {
+        override def run(): Unit = {
+          var shouldRun = true
+          while (shouldRun) {
+            semaphore.acquire()
+
+            if (lastValue == null) {
+              shouldRun = false
+            } else {
+              lastValue.foreach { v =>
+                ptr.get.foreach(_.publishValue(v))
+              }
+            }
+          }
+        }
+      })
+
+      thread.start()
+
+
+      var cancel: Cancel = null
+      cancel = foreach { v =>
+        ptr.get match {
+          case Some(s) =>
+            lastValue = Some(v)
+            semaphore.release()
+
+          case None =>
+            lastValue = null
+            semaphore.release()
+        }
+      }
+
+      ret
+    } else {
+      this
+    }
+  }
+
+  /**
+    * Creates a stream that emits values at the given rate by polling
+    * from the original stream
+    * @param period the period to poll at
+    * @param clock the clock to use to schedule periodic events
+    * @return a stream emitting values from the original stream at the new rate
+    */
+  def pollPeriodic(period: Time)(implicit clock: Clock): Stream[T] = {
+    syncTo(Stream.periodic(period)(()))
+  }
+
+  /**
+    * Adds a "check" in the stream pipeline, which can be used for things
+    * such as ending a task when a condition is met
+    * @param check the function to run on each value of the stream
+    * @return a stream that contains the check in the pipeline
+    */
+  def withCheck(check: T => Unit): Stream[T] = {
+    map { v =>
+      check(v)
+      v
+    }
+  }
+
+  /**
+    * Filters a stream to only emit values that pass a certain condition
+    * @param condition the condition to filter stream values with
+    * @return a stream that only emits values that pass the condition
+    */
+  def filter(condition: T => Boolean): Stream[T] = {
+    val ret = new Stream[T] {
+      val parent = self
+      override val expectedPeriodicity = NonPeriodic
+    }
+
+    val ptr = WeakReference(ret)
+
+    var cancel: Cancel = null
+    cancel = this.foreach { v =>
+      ptr.get match {
+        case Some(s) =>
+          if (condition(v)) {
+            s.publishValue(v)
+          }
+        case None =>
+          cancel()
+      }
+    }
+
+    ret
+  }
+
+  /**
+    * Adds a listener for elements of this stream. Callbacks will be executed
+    * whenever a new value is published in order of when the callbacks were added.
+    * Callbacks added first will be called first and callbacks added last will
+    * be called last.
+    *
+    * @param thunk the listener to be called on each published value
+    * @return a function to call to remove the listener
+    */
+  def foreach(thunk: T => Unit): Cancel = {
+    synchronized {
+      listeners.enqueue(thunk)
```

Comment:
Yeah, foreach takes a function to be called with each value emitted by the stream.